### DISCUSSION TITLE
Optional Path(...) parameter

### DISCUSSION THREAD

**Discussion by LKay** [fastapi/fastapi#8067] [Questions]
**Description**

Let's assume I have two routes:
 * `/notes`
 * `/users/{user_id}/notes`

Both will return the same format of the response but the route prefixed by `/users/{user_id}` will be additionally filtering notes by specific `user_id`.

I would like to not duplicate the logic unnecessarily and have something like function below to parse parameters:

```python
async def params(user_id: Optional[str] = Path(None)) -> Optional[str]:
    return user_id

@api.get("/notes")
async def root_get(user_id: Optional[str] = Depends(params)):
    pass

@api.get("/users/{user_id}/notes")
async def user_get(user_id: Optional[str] = Depends(params)):
    pass
```

However `Path` only accepts defined `str` and doesn't allow `None`.

Is there any way to make path parameters optional?

**Comment by jonathanunderwood** [comment]
This is what I've implemented in PR #464 and raised in #450. 

Waiting on @tiangolo's feedback on #464.

**Comment by jonathanunderwood** [comment]
#464 has now been merged, so I think this can be closed.

**Comment by tiangolo** [comment]
Thanks for the request @LKay.

This was implemented by @jonathanunderwood and is available in the latest versions. 🚀

I'll close this issue now, but feel free to add more comments or create new issues.

**Comment by requiemofthesouls** [comment]
seems I have similar question.

**Description:**
I have a function which connected with two routes:

- `/{device}/mtu`
- `/{device}/{interface}/mtu`

```python3
@router.get("/{device}/mtu")
async def ifMtuAll(device: str, bg_task: BackgroundTasks):
    return await cached_snmp_with_port_filtering(device, bg_task, func="ifMtu")


@router.get("/{device}/{interface}/mtu")
async def ifMtuExact(device: str, bg_task: BackgroundTasks, interface: int):
    return await cached_snmp_with_port_filtering(device, bg_task, func="ifMtu", interface=interface)
```
I want parameter `device` should be obligatory, but `interface` shouldn't.
Is it possible to make `interface` parameter optional and get rid of two routes to one?
I tried this solution with `Path(None)`, but I get 404 error when don't filling `interface` parameter.  If I try this in swagger it substitute name of that parameter: `http://localhost:8084/ports/myswitch/`**{interface}**`/mtu`
```python3
async def params(interface: Optional[int] = Path(None)) -> Optional[int]:
    return interface

@router.get("/{device}/{interface}/mtu")
async def ifMtu(device: str, bg_task: BackgroundTasks, interface: Optional[int] = Depends(params)):
    return await cached_snmp_with_port_filtering(device, bg_task, func="ifMtu", interface=interface)
```
And examples:
It works with `interface` parameter
```bash
$ curl -X GET "http://localhost:8084/ports/myswitch/2/mtu" -H "accept: application/json"  
{"DEVICE":"myswitch","VALUE":{"2":"1500"}}
```
Not found without it
```bash
$ curl -X GET "http://localhost:8084/ports/myswitch/mtu" -H "accept: application/json"  
{"detail":"Not Found"}
```
And example in swagger with empty `interface` parameter.
```bash
$ curl -X GET "http://localhost:8084/ports/bs90678/{interface}/mtu" -H "accept: application/json" 
{"detail":[{"loc":["path","interface"],"msg":"value is not a valid integer","type":"type_error.integer"},{"loc":["path","interface"],"msg":"value is not none","type":"type_error.none.allowed"}]}
```

**Comment by tiangolo** [comment]
@requiemofthesouls if you are still having problems please create a new issue with a minimal self-contained example, that I (or someone else) can try it out and help you better.
